### ISSUE TITLE
Learning the structure of generative models without labeled data

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#365]
Learning the structure of generative models without labeled data<br>
<br>
https://ift.tt/2wloISy<br>
<br>
adriancolyer<br>
<br>
<div>
<div class="entry clear">
<p><a href="https://arxiv.org/abs/1703.00854">Learning the structure of generative models without labeled data</a> Bach et al., <em>ICML&rsquo;17</em></p>
<p>For the last couple of posts we&rsquo;ve been looking at <a href="https://blog.acolyer.org/2018/08/22/snorkel-rapid-training-data-creation-with-weak-supervision">Snorkel</a> and <a href="https://blog.acolyer.org/2018/08/24/training-classifiers-with-natural-language-explanations">BabbleLabble</a> which both depend on <em>data programming</em> &ndash; the ability to intelligently combine the outputs of a set of labelling functions. The core of data programming is developed in two papers, &lsquo;<a href="https://arxiv.org/abs/1605.07723">Data programming: creating large training sets, quickly</a>&rsquo; (Ratner 2016) and today&rsquo;s paper choice, &lsquo;<a href="https://arxiv.org/abs/1703.00854">Learning the structure of generative models without labeled data</a>&rsquo; (Bach 2017).</p>
<p>The original data programming paper works explicitly with input pairs (x,y) (e.g. the chemical and disease word pairs we saw from the disease task in Snorkel) which (for me at least) confuses the presentation a little compared to the latter ICML paper which just assumes inputs <img src="https://s0.wp.com/latex.php?latex=x&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="x" title="x" class="latex"> (which could of course have pair structure, but we don&rsquo;t care about that at this level of detail). Also in the original paper dependencies between labelling functions are explicitly specified by end users (as one of four types: similar, fixing, reinforcing, and exclusive) and built into a factor graph. In the ICML paper dependencies are learned. So I&rsquo;m going to work mostly from &lsquo;Learning the structure of generative models without labeled data&rsquo; in this post, and concentrate just on the core development of the theory.</p>
<h3>A generative probabilistic model for accuracy estimation under independence</h3>
<p>We have input data points <img src="https://s0.wp.com/latex.php?latex=x_%7Bi%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="x_{i}" title="x_{i}" class="latex">, and for each <img src="https://s0.wp.com/latex.php?latex=x_%7Bi%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="x_{i}" title="x_{i}" class="latex"> a latent variable <img src="https://s0.wp.com/latex.php?latex=y+%5Cin+%5C%7B-1%2C+1%5C%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="y \in \{-1, 1\}" title="y \in \{-1, 1\}" class="latex"> that is its true label. We don&rsquo;t know <img src="https://s0.wp.com/latex.php?latex=y_%7Bi%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="y_{i}" title="y_{i}" class="latex">, but we do have <img src="https://s0.wp.com/latex.php?latex=n&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="n" title="n" class="latex"> labelling functions <img src="https://s0.wp.com/latex.php?latex=%5Clambda_%7B1%7D%2C+...%2C+%5Clambda_%7Bn%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\lambda_{1}, ..., \lambda_{n}" title="\lambda_{1}, ..., \lambda_{n}" class="latex">.</p>
<p>Each labelling functions takes as input an data point <img src="https://s0.wp.com/latex.php?latex=x_i&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="x_i" title="x_i" class="latex"> and produces as output a value <img src="https://s0.wp.com/latex.php?latex=%5CLambda&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\Lambda" title="\Lambda" class="latex"> which is either true, abstain, or false ( <img src="https://s0.wp.com/latex.php?latex=%5C%7B1%2C+0%2C+-1%5C%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\{1, 0, -1\}" title="\{1, 0, -1\}" class="latex"> ). (The approach generalises to the multi-class case).</p>
<p>We start out by assuming that the outputs of the labelling functions are conditionally <em>independent</em> given the true label. Consider a labelling function <img src="https://s0.wp.com/latex.php?latex=%5Clambda_%7Bj%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\lambda_{j}" title="\lambda_{j}" class="latex">, which gives output <img src="https://s0.wp.com/latex.php?latex=%5CLambda_%7Bij%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\Lambda_{ij}" title="\Lambda_{ij}" class="latex"> on input <img src="https://s0.wp.com/latex.php?latex=x_%7Bi%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="x_{i}" title="x_{i}" class="latex"> (with true label <img src="https://s0.wp.com/latex.php?latex=y_i&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="y_i" title="y_i" class="latex">). We can model the <em>accuracy</em> of this labelling function with <img src="https://s0.wp.com/latex.php?latex=%5Cphi_%7Bj%7D%5E%7BAcc%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\phi_{j}^{Acc}" title="\phi_{j}^{Acc}" class="latex">. We&rsquo;ll say the accuracy function has value 1 if <img src="https://s0.wp.com/latex.php?latex=%5Clambda_%7Bj%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\lambda_{j}" title="\lambda_{j}" class="latex"> correctly labels an example, and 0 otherwise. In our setting this can be succinctly expressed as:</p>
<p><img src="https://s0.wp.com/latex.php?latex=%5Cdisplaystyle+%5Cphi_%7Bj%7D%5E%7BAcc%7D%28%5CLambda_i%2C+y_i%29+%3D+y_i+%5CLambda_%7Bij%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\displaystyle \phi_{j}^{Acc}(\Lambda_i, y_i) = y_i \Lambda_{ij}" title="\displaystyle \phi_{j}^{Acc}(\Lambda_i, y_i) = y_i \Lambda_{ij}" class="latex"></p>
<p>The accuracy function itself is controlled by some parameter <img src="https://s0.wp.com/latex.php?latex=%5Ctheta_%7Bj%7D%5E%7BAcc%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\theta_{j}^{Acc}" title="\theta_{j}^{Acc}" class="latex"> modelling how accurate each labelling function is.</p>
<p>Now for <img src="https://s0.wp.com/latex.php?latex=m&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="m" title="m" class="latex"> input examples and <img src="https://s0.wp.com/latex.php?latex=n&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="n" title="n" class="latex"> labelling functions, we have a conditionally independent model</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/learning-structure-eqn-1.jpeg?w=480"></p>
<p>where <img src="https://s0.wp.com/latex.php?latex=Y&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="Y" title="Y" class="latex"> is the set of true labels <img src="https://s0.wp.com/latex.php?latex=y_1%2C+...%2C+y_m&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="y_1, ..., y_m" title="y_1, ..., y_m" class="latex">.</p>
<p>The unknown parameters <img src="https://s0.wp.com/latex.php?latex=%5Ctheta&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\theta" title="\theta" class="latex"> modelling the accuracy of the labelling functions can now be estimated by minimising the negative log marginal likelihood of failing to give the correct label (guessing a wrong label, or abstaining).</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/learning-structure-eqn-2.jpeg?w=480"></p>
<p>We can do this using standard stochastic gradient descent, where the gradient for parameter <img src="https://s0.wp.com/latex.php?latex=%5Ctheta_%7Bj%7D%5E%7BAcc%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\theta_{j}^{Acc}" title="\theta_{j}^{Acc}" class="latex"> is given by the number of examples <img src="https://s0.wp.com/latex.php?latex=%5Clambda_j&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\lambda_j" title="\lambda_j" class="latex"> correctly labels minus the number of examples it fails to label correctly.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/learning-structure-i-gradient.jpeg?w=520"></p>
<h3>Structure learning</h3>
<blockquote>
<p>The conditionally independent model is a common assumption, and using a more sophisticated generative model currently requires users to specify its structure.</p>
</blockquote>
<p>As we know from the Snorkel paper, unfortunately labelling functions are often not independent.</p>
<blockquote>
<p>To address this issue, we generalize the conditionally independent model as a factor graph with additional dependencies, including higher-order factors that connect multiple labeling function outputs for each data point <img src="https://s0.wp.com/latex.php?latex=x_i&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="x_i" title="x_i" class="latex"> and label <img src="https://s0.wp.com/latex.php?latex=y_i&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="y_i" title="y_i" class="latex">.</p>
</blockquote>
<p>Suppose there is a set <img src="https://s0.wp.com/latex.php?latex=T&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="T" title="T" class="latex"> of dependency types of interest, where a dependency type represents a form of dependency between labelling functions, such as correlation. Unknown to us, there is also a set <img src="https://s0.wp.com/latex.php?latex=S_t&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="S_t" title="S_t" class="latex"> of index tuples which indicate the set of labelling functions that participate in each dependency of type <img src="https://s0.wp.com/latex.php?latex=t+%5Cin+T&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="t \in T" title="t \in T" class="latex">. This gives us a general model of the form</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/structure-learning-eqn-3.jpeg?w=480"></p>
<p>and we want to learn the parameters <img src="https://s0.wp.com/latex.php?latex=%5Ctheta_%7Bs%7D%5E%7Bt%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\theta_{s}^{t}" title="\theta_{s}^{t}" class="latex">.</p>
<blockquote>
<p>Estimating the structure of the distribution <img src="https://s0.wp.com/latex.php?latex=p_%7B%5Ctheta%7D%28%5CLambda%2C+Y%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="p_{\theta}(\Lambda, Y)" title="p_{\theta}(\Lambda, Y)" class="latex"> is challenging because Y is latent; we never observe its value, even during training. We must therefore work with the marginal likelihood <img src="https://s0.wp.com/latex.php?latex=p_%7B%5Ctheta%7D%28%5CLambda%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="p_{\theta}(\Lambda)" title="p_{\theta}(\Lambda)" class="latex">. Learning the parameters of the generative model requires Gibbs sampling to estimate gradients. As the number of possible dependencies increases at least quadratically in the number of labeling functions, this heavyweight approach to learning does not scale.</p>
</blockquote>
<p>The way out is to consider each labelling function in turn, optimising the log marginal pseudolikelihood of its outputs conditioned on the outputs of all the other labelling functions.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/structure-learning-eqn-4.jpeg?w=480"></p>
<p>where <img src="https://s0.wp.com/latex.php?latex=%5Cepsilon&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\epsilon" title="\epsilon" class="latex"> is a hyperparameter.</p>
<blockquote>
<p>By conditioning on all other labeling functions in each term&hellip;, we ensue that the gradient can be computed in polynomial time with respect to the number of labeling functions, data points, and possible dependencies; without requiring any sampling or variational approximations.</p>
</blockquote>
<p>It all comes together in the following algorithm:</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/structure-learning-alg-1.jpeg?w=480"></p>
<p>The hyperparameter <img src="https://s0.wp.com/latex.php?latex=%5Cepsilon&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\epsilon" title="\epsilon" class="latex"> controls the threshold and regularization strength and requires tuning. For the other parameters the authors fix these at step size = <img src="https://s0.wp.com/latex.php?latex=m%5E%7B-1%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="m^{-1}" title="m^{-1}" class="latex">, epoch count = 10, and truncation frequency = 10.</p>
<h3>Theoretical guarantees</h3>
<p>Assuming that there exist some set of parameters which can capture the true dependencies within the model we are using, and that all non-zero parameters are bounded away from zero (have at least a minimum magnitude <img src="https://s0.wp.com/latex.php?latex=%5Ckappa&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\kappa" title="\kappa" class="latex">), then we can uncover the exact dependency structure with probability at least <img src="https://s0.wp.com/latex.php?latex=1+-+%5Cdelta&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="1 - \delta" title="1 - \delta" class="latex"> given <img src="https://s0.wp.com/latex.php?latex=m&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="m" title="m" class="latex"> inputs, where</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/structure-learning-theorem-1.jpeg?w=320"></p>
<p>Where <img src="https://s0.wp.com/latex.php?latex=d&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="d" title="d" class="latex"> is the maximum number of possible dependencies a single labelling function (or true label) can be involved in, and <img src="https://s0.wp.com/latex.php?latex=c&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="c" title="c" class="latex"> is a measure of how much better our dependency estimates are with each labeling function than without (my loose interpretation of <img src="https://s0.wp.com/latex.php?latex=c&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="c" title="c" class="latex"> &ndash; see eqn 7 in section 4 of the paper for the official definition).</p>
<p>If we further assume that the only dependency types are accuracy and correlation dependencies then we need an input dataset of size</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/structure-learning-theorem-2.jpeg?w=320"></p>
<h3>Practical results</h3>
<blockquote>
<p>Our analysis&hellip; guarantees that the sample complexity grows at worst on the order <img src="https://s0.wp.com/latex.php?latex=O%28n+%5Clog+n%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="O(n \log n)" title="O(n \log n)" class="latex"> for <img src="https://s0.wp.com/latex.php?latex=n&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="n" title="n" class="latex"> labeling functions. In practice, we find that structure learning performs better than this guaranteed rate, depending linearly on the number of true correlations and logarithmically on the number of possible correlations.</p>
</blockquote>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/structure-learning-fig-2.jpeg?w=480"></p>
<p>Compared to estimating structures via parameter learning over all possible dependencies, algorithm one is 100x faster and selects only 1/4 of the extraneous correlations.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/structure-learning-fig-3.jpeg?w=480"></p>
</div>
</div>
<br>
<br>
<br>
<br>
BigData,AI|ML|DM|IR|PR<br>
<br>
via the morning paper https://ift.tt/1WJ1E4W<br>
<br>
August 28, 2018 at 01:00PM
